sysdig/resource_sysdig_secure_cloud_auth_account.go

```diff
@@ -135,39 +236,211 @@ func resourceSysdigSecureCloudauthAccountDelete(ctx context.Context, data *schem
+			for key, value := range entry {
+				result[key] = value
+			}
+		}
+	}
+
+	return result
+}
+
 func cloudauthAccountFromResourceData(data *schema.ResourceData) *v2.CloudauthAccountSecure {
+	components := []*cloudauth.AccountComponent{}
+
+	for _, rc := range data.Get("components").([]interface{}) {
+		resourceComponent := rc.(map[string]interface{})
+		component := &cloudauth.AccountComponent{}
+
+		for key, value := range resourceComponent {
+
+			if value != nil && value.(string) != "" {
+				switch key {
+				case "type":
+					component.Type = cloudauth.Component(cloudauth.Component_value[value.(string)])
+				case "instance":
+					component.Instance = value.(string)
+				case "cloud_connector_metadata":
+					component.Metadata = &cloudauth.AccountComponent_CloudConnectorMetadata{
+						CloudConnectorMetadata: &cloudauth.CloudConnectorMetadata{},
+					}
+				case "trusted_role_metadata":
+					// TODO: Make it more generic than just for GCP
+					component.Metadata = &cloudauth.AccountComponent_TrustedRoleMetadata{
+						TrustedRoleMetadata: &cloudauth.TrustedRoleMetadata{
+							Provider: &cloudauth.TrustedRoleMetadata_Gcp{
+								Gcp: &cloudauth.TrustedRoleMetadata_GCP{
+									RoleName: value.(string),
+								},
+							},
+						},
+					}
+				case "event_bridge_metadata":
+					component.Metadata = &cloudauth.AccountComponent_EventBridgeMetadata{
+						EventBridgeMetadata: &cloudauth.EventBridgeMetadata{},
+					}
+				case "service_principal_metadata":
+					// TODO: Make it more generic than just for GCP
+					component.Metadata = &cloudauth.AccountComponent_ServicePrincipalMetadata{
+						ServicePrincipalMetadata: &cloudauth.ServicePrincipalMetadata{
+							Provider: &cloudauth.ServicePrincipalMetadata_Gcp{
+								Gcp: &cloudauth.ServicePrincipalMetadata_GCP{
+									Key: &cloudauth.ServicePrincipalMetadata_GCP_Key{
+										ProjectId:    data.Get("cloud_provider_id").(string),
+										PrivateKeyId: "deadbeef",
+										PrivateKey:   "cert thangs",
+									},
+								},
+							},
+						},
+					}
+				case "webhook_datasource_metadata":
+					component.Metadata = &cloudauth.AccountComponent_WebhookDatasourceMetadata{
+						WebhookDatasourceMetadata: &cloudauth.WebhookDatasourceMetadata{},
+					}
+				case "crypto_key_metadata":
+					component.Metadata = &cloudauth.AccountComponent_CryptoKeyMetadata{
+						CryptoKeyMetadata: &cloudauth.CryptoKeyMetadata{},
+					}
+				case "cloud_logs_metadata":
+					component.Metadata = &cloudauth.AccountComponent_CloudLogsMetadata{
+						CloudLogsMetadata: &cloudauth.CloudLogsMetadata{},
+					}
+				}
+			}
+		}
+
+		components = append(components, component)
+	}
+
+	accountFeatures := &cloudauth.AccountFeatures{}
+
+	featureData := data.Get("feature")
+
+	var featureMap map[string]interface{}
+
+	featureMap = convertSchemaSetToMap(featureData.(*schema.Set))
+
+	for name, value := range featureMap {
```

Comment:
most of the cases have similar implementation, can we try to refactor and make it more readable?